speedyfx/lib/speedyfx.c

```diff
@@ -0,0 +1,65 @@
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <ctype.h>
+#include <err.h>
+#include <fcntl.h>
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+/*
+ * SpeedyFx algorithm
+ * ==================
+ *
+ * Tokenize/hash large amount of strings efficiently.
+ * Original Java implementation: http://www.hpl.hp.com/techreports/2008/HPL-2008-91R1.pdf
+ * Ported to C by Stanislaw Pusep (https://github.com/creaktive)
+ * Compile with:
+ *      clang -lm -O3 -o speedyfx speedyfx.c
+ * or:
+ *      gcc -lm -O3 -o speedyfx speedyfx.c
+ * Then use as:
+ *      ./speedyfx enwik9 > fv.bin
+ * To generate 128KB feature vector for enwik9 text file.
+ *
+ * Benchmark
+ * =========
+ *
+ * Test data: https://cs.fit.edu/~mmahoney/compression/enwik9.bz2
+ * Hardware: Intel(R) Xeon(R) CPU E5620 @ 2.40GHz
+ * Average feature vector build speed: 213.83 MB/s
+ */
+
+#define SetBit(a, b) (((unsigned char *) a)[(b) >> 3] |= (1 << ((b) & 7)))
+
+unsigned char *speedyfx_fv(const unsigned char *s, unsigned int *code_table,
+                                        unsigned int n, unsigned int length) {
+    unsigned int code, c;
+    unsigned int wordhash = 0;
+    unsigned char *fv;
+
+    fv = calloc(ceil((float) n / 8.0), sizeof(unsigned char));
```

Comment:
So this is not very optimal, essentially we are going to store a few value into an array where most elements are going to be `0`, then to transform it to a sparse scipy array, we will have to scann through all the elements and only take the ones that are  non zero.